invoices/invoiceregistry.go

```diff
@@ -300,61 +300,25 @@ func (i *InvoiceRegistry) invoiceEventLoop() {
 		}
 
 		select {
-		// A new invoice subscription for all invoices has just arrived!
-		// We'll query for any backlog notifications, then add it to the
-		// set of clients.
-		case newClient := <-i.newSubscriptions:
-			log.Infof("New invoice subscription "+
-				"client: id=%v", newClient.id)
-
-			// With the backlog notifications delivered (if any),
-			// we'll add this to our active subscriptions and
-			// continue.
-			i.notificationClients[newClient.id] = newClient
-
 		// A client no longer wishes to receive invoice notifications.
 		// So we'll remove them from the set of active clients.
 		case clientID := <-i.subscriptionCancels:
 			log.Infof("Cancelling invoice subscription for "+
 				"client=%v", clientID)
+			i.deleteClient(clientID)
```

Comment:
Can we still run into the issue where we cancel before we've added a client here? 

An alternative here would be to add a `subscriptionModifications` channel with a new interface (mostly used as a marker interface): 
```
type subscriptionModification interface{
   func invoiceSubscriptionID() uint32
}
```

Subscription additions (`subscriptionModificationAdd`) and cancellations (`subscritionModificationCancel`) could implement this interface, and send events into the `InvoiceRegistry` single modifications channel. 

Since a subscription has to be read from the channel (and thus fully handled) before the caller has access to the `Cancel` function. This is where I think that the single select to handle all cases pattern (used in `invoiceEventLoop`) shines.

Reply:
> Can we still run into the issue where we cancel before we've added a client here?

We won't, as the client is not returned unless it's registered first.

> Since a subscription has to be read from the channel (and thus fully handled) before the caller has access to the Cancel function.

So how's to prevent `Cancel` being called before registration in this case? We still can't guarantee the order of the  add event and cancel event? Or we can use a unbuffered channel and block until that add event is processed?

Reply:
> So how's to prevent Cancel being called before registration in this case?

If they're both being sent into the same channel, we always send `Add` before `Cancel` can be sent, so they'll only ever be processed in order (since we send the `Add` event into the channel, then return and `Cancel` can be called). Don't need to be blocking, can have a buffered channel and the events just queue themselves up in order.

Reply:
I see, same channel to act as a queue. As long as `cancel` is before `add` we'd be good no mater how many clients are registered. Neat!

Reply:
Tried [this alternative approach](https://github.com/lightningnetwork/lnd/compare/master...yyforyongyu:alter-itest-fix-invoice-registry?expand=1#diff-1e2a471f2e4e390c0573ac8b818696ffb77f5c44ac04eb64049854cc3957df43R1552), the channel communication looks good, the only downside is the long and repetitive check on `i.quit`. I think we could refactor the registration and even-delivering logic to get rid of that. As for now, it seems easier to just add and delete inside locks.